exonum/src/api/backends/actix.rs

```diff
@@ -131,116 +132,143 @@ impl ResponseError for api::Error {
+/// "<warn-code> <warn-agent> \"<warn-text>\" [<warn-date>]"
+/// <warn-code> in our case is 299, which means a miscellaneous persistent warning.
+/// <warn-agent> is optional, so we set it to "-".
+/// <warn-text> is a warning description, which is taken as an only argument.
+/// <warn-date> is not required.
+/// For details you can see RFC 7234, section 5.5: Warning.
+fn create_warning_header(warning_text: &str) -> String {
+    const WARNING_NUMBER: u16 = 299;
+    const WARNING_AGENT: &str = "-";
+
+    format!("{} {} \"{}\"", WARNING_NUMBER, WARNING_AGENT, warning_text)
+}
+
+impl From<EndpointMutability> for actix_web::http::Method {
+    fn from(mutability: EndpointMutability) -> Self {
+        match mutability {
+            EndpointMutability::Immutable => actix_web::http::Method::GET,
+            EndpointMutability::Mutable => actix_web::http::Method::POST,
         }
     }
 }
 
-impl<Q, I, F> From<NamedWith<Q, I, api::Result<I>, F, Mutable>> for RequestHandler
+impl<Q, I, F> From<NamedWith<Q, I, api::Result<I>, F>> for RequestHandler
 where
     F: Fn(Q) -> api::Result<I> + 'static + Send + Sync + Clone,
     Q: DeserializeOwned + 'static,
     I: Serialize + 'static,
 {
-    fn from(f: NamedWith<Q, I, api::Result<I>, F, Mutable>) -> Self {
+    fn from(f: NamedWith<Q, I, api::Result<I>, F>) -> Self {
+        // Convert handler that returns a `Result` into handler that will return `FutureResult`.
         let handler = f.inner.handler;
-        let index = move |request: HttpRequest| -> FutureResponse {
-            let handler = handler.clone();
-            request
-                .json()
-                .from_err()
-                .and_then(move |query: Q| {
-                    handler(query)
-                        .map(|value| HttpResponse::Ok().json(value))
-                        .map_err(From::from)
-                })
-                .responder()
+        let future_endpoint = move |query| -> Box<dyn Future<Item = I, Error = api::Error>> {
+            let future = handler(query).into_future();
+            Box::new(future)
         };
+        let named_with_future = NamedWith::new(f.name, future_endpoint, f.mutability);
 
-        Self {
-            name: f.name,
-            method: actix_web::http::Method::POST,
-            inner: Arc::from(index) as Arc<RawHandler>,
-        }
+        // Then we can create a `RequestHandler` with the `From` specialization for future result.
+        RequestHandler::from(named_with_future)
     }
 }
 
-impl<Q, I, F> From<NamedWith<Q, I, FutureResult<I>, F, Immutable>> for RequestHandler
+/// Takes `HttpRequest` as a parameter and extracts query:
+/// - If request is immutable, the query is parsed from query string,
+/// - If request is mutable, the query is parsed from the request body as JSON.
+fn extract_query<Q>(
+    request: HttpRequest,
+    mutability: EndpointMutability,
+) -> Box<dyn Future<Item = Q, Error = actix_web::error::Error>>
```

Comment:
Nit: Couldn't this be `impl Future`?

Reply:
Unfortunately no, match arms then will have incompatible types.

Reply:
And what about Either?

Reply:
`Either` will make the return type more complex. As far as I know `Box` doesn't bring any overhead other than memory allocation, and since we aren't afraid of creating `Vec`s, this isn't a big problem, is it?

Reply:
> `Either` will make the return type more complex.

How though? The return type will be `impl Future<..>`.

> As far as I know Box doesn't bring any overhead other than memory allocation

Well, memory allocation in heap *is* an overhead, so we use arrays where it's possible (e.g., in crypto). Additionally, `Box<dyn _>` brings in dynamic dispatch, which is also slower. I doubt this overhead is significant, but if we *can* use stack-enabled / statically dispatched `impl Future`, why not?